File: src/core/following.ts
import type { DriverBehaviour, FollowingMode, VehicleType } from './types';

export interface Leader { readonly gap: number; readonly speed: number }
export interface FollowingResult { readonly acceleration: number; readonly mode: FollowingMode }

/**
 * Reduced four-regime, Wiedemann-inspired prototype. This is NOT W74 or W99.
 * Safety-distance shape follows W74; thresholds and acceleration laws are our own.
 * See docs/SIMULATION.md for equations, limitations and the source reference.
 */
export function followingAcceleration(
  speed: number, desiredSpeed: number, driverFactor: number,
  type: VehicleType, behaviour: DriverBehaviour, leader?: Leader,
): FollowingResult {
  const free = Math.min(type.maxAcceleration, (desiredSpeed - speed) / behaviour.followingTime);
  let acceleration = free;
  let mode: FollowingMode = 'free';
  if (leader) {
    const desiredGap = behaviour.standstillDistance +
      (behaviour.additiveSafetyDistance + behaviour.multiplicativeSafetyDistance * driverFactor) *
      Math.sqrt(Math.max(0, speed));
    const closing = speed - leader.speed;
    const room = Math.max(0.01, leader.gap - desiredGap);
    if (leader.gap < desiredGap) {
      mode = 'braking';
      acceleration = Math.min(free, -closing / behaviour.followingTime +
        (leader.gap - desiredGap) / behaviour.followingTime ** 2);
    } else if (closing > behaviour.speedThreshold &&
        room < closing ** 2 / (2 * type.comfortableDeceleration) + speed * behaviour.followingTime) {
      mode = 'approaching';
      acceleration = Math.min(free, -(closing ** 2) / (2 * room));
    } else if (leader.gap < desiredGap * 1.5) {
      mode = 'following';
      acceleration = Math.min(free, -closing / behaviour.followingTime +
        (leader.gap - desiredGap) / behaviour.followingTime ** 2);
    }
  }
  return { acceleration: Math.max(-type.maxDeceleration, acceleration), mode };
}

/** Ballistic displacement with a within-step stop; never integrates backwards. */
export function integrate(speed: number, acceleration: number, dt: number): { speed: number; distance: number } {
  if (acceleration < 0 && speed + acceleration * dt < 0) {
    return { speed: 0, distance: -(speed ** 2) / (2 * acceleration) };
  }
  return { speed: Math.max(0, speed + acceleration * dt), distance: Math.max(0, speed * dt + 0.5 * acceleration * dt ** 2) };
}

File: src/core/following.cpp
#include "following.hpp"
#include <algorithm>
#include <cmath>

namespace trafficsim {
// Reduced Wiedemann-inspired prototype, NOT W74/W99. See docs/SIMULATION.md.
FollowingResult followingAcceleration(double speed, double desiredSpeed, double driverFactor,
    const VehicleType& type, const DriverBehaviour& behaviour, std::optional<Leader> leader) {
    const double free = std::min(type.maxAcceleration, (desiredSpeed - speed) / behaviour.followingTime);
    double acceleration = free;
    auto mode = FollowingMode::free;
    if (leader) {
        const double desiredGap = behaviour.standstillDistance +
            (behaviour.additiveSafetyDistance + behaviour.multiplicativeSafetyDistance * driverFactor) *
            std::sqrt(std::max(0.0, speed));
        const double closing = speed - leader->speed;
        const double room = std::max(0.01, leader->gap - desiredGap);
        const auto following = [&] {
            return std::min(free, -closing / behaviour.followingTime + (leader->gap - desiredGap) /
                            (behaviour.followingTime * behaviour.followingTime));
        };
        if (leader->gap < desiredGap) {
            mode = FollowingMode::braking;
            acceleration = following();
        } else if (closing > behaviour.speedThreshold &&
                   room < closing * closing / (2 * type.comfortableDeceleration) +
                          speed * behaviour.followingTime) {
            mode = FollowingMode::approaching;
            acceleration = std::min(free, -(closing * closing) / (2 * room));
        } else if (leader->gap < desiredGap * 1.5) {
            mode = FollowingMode::following;
            acceleration = following();
        }
    }
    return {std::max(-type.maxDeceleration, acceleration), mode};
}
Motion integrate(double speed, double acceleration, double dt) {
    if (acceleration < 0 && speed + acceleration * dt < 0)
        return {0, -(speed * speed) / (2 * acceleration)};
    return {std::max(0.0, speed + acceleration * dt),
            std::max(0.0, speed * dt + 0.5 * acceleration * (dt * dt))};
}
}
